include/sparrow/array/external_array_data.hpp

```diff
@@ -133,73 +164,77 @@ namespace sparrow
 
     namespace impl
     {
-        template <class T>
+        template <any_arrow_c_interface T>
         template <class U>
-        requires std::is_rvalue_reference_v<U&&>
-        external_wrapper<T>::external_wrapper(U&& u, bool own)
+            requires std::convertible_to<U, T>
+        external_wrapper<T>::external_wrapper(U&& u, ownership ownership_model)
         {
-            if (own)
+            if (ownership_model == ownership::owning)
             {
-                auto l = [r = releaser()](T* t)
-                {
-                    r(t);
-                    delete t;
-                };
-                p_data = std::shared_ptr<T>(new T(std::move(u)), std::move(l));
+                p_data = std::shared_ptr<T>(new T(std::forward<U>(u)), full_ownership_deleter{});
             }
             else
             {
-                p_data = std::shared_ptr<T>(new T(std::move(u)));
+                p_data = std::shared_ptr<T>(new T(std::forward<U>(u))); // default shared_ptr deleter
             }
         }
 
-        template <class T>
+        template <any_arrow_c_interface T>
         template <class U>
-        requires std::is_pointer_v<U>
-        external_wrapper<T>::external_wrapper(U u, bool own)
+            requires std::is_pointer_v<std::remove_reference_t<U>> and std::convertible_to<U, T*>
+        external_wrapper<T>::external_wrapper(U ptr, ownership ownership_model)
         {
-            if (own)
+            if (ownership_model == ownership::owning)
             {
-                p_data = std::shared_ptr<T>(u, releaser());
+                p_data = std::shared_ptr<T>(ptr, only_release_arrow_data{});
             }
             else
             {
-                p_data = std::shared_ptr<T>(u, no_releaser());
+                p_data = std::shared_ptr<T>(ptr, dont_release_anything{});
             }
         }
 
-        template <class T>
+        template <any_arrow_c_interface T>
         T& external_wrapper<T>::data()
         {
             return *p_data;
         }
 
-        template <class T>
+        template <any_arrow_c_interface T>
         const T& external_wrapper<T>::data() const
         {
             return *p_data;
         }
 
-        template <class T>
-        void external_wrapper<T>::releaser::operator()(T* t) const
+        template <any_arrow_c_interface T>
+        void external_wrapper<T>::only_release_arrow_data::operator()(T* t) const
         {
             if (t->release)
             {
                 t->release(t);
                 t->release = nullptr;
             }
         }
+
+        template <any_arrow_c_interface T>
+        void external_wrapper<T>::full_ownership_deleter::operator()(T* t) const
+        {
+            if (t->release)
+            {
+                t->release(t);
+            }
```

Comment:
you should also set the release member to nullptr

Reply:
Is that necessary when we do `delete t` the next line?

Reply:
Forgot that, the release function should set the t->release = NULL: https://arrow.apache.org/docs/dev/format/CDataInterface.html#release-callback-semantics-for-producers
It's the case in our release function implementation.